primitives/src/balances.rs

```diff
@@ -56,6 +61,89 @@ pub struct BalanceEntry<BlockNumber> {
 	pub last_update: BlockNumber,
 }
 
+impl<BlockNumber> BalanceEntry<BlockNumber>
+where
+	BlockNumber: AtLeast32Bit + Debug,
+{
+	pub fn new(principal: BalanceType, last_update: BlockNumber) -> Self {
+		Self { principal, last_update }
+	}
+
+	/// Applies the demurrage and returns an updated BalanceEntry.
+	///
+	/// The following formula is applied to the principal:
+	///    updated_principal = old_principal * e^(-demurrage_per_block * elapsed_blocks)
+	///
+	/// If the `demurrage_per_block` is negative it will take the absolute value of
+	/// it for the calculation.
+	///
+	/// **Note**: This function will be used at every single transaction that is paid with community
+	/// currency. It is important that it is as efficient as possible, but also that it is bullet-
+	/// proof (no-hidden panics!!).
+	pub fn apply_demurrage(
+		self,
+		demurrage_per_block: Demurrage,
+		current_block_number: BlockNumber,
+	) -> BalanceEntry<BlockNumber> {
+		if self.last_update == current_block_number {
+			// Nothing to be done, as no time elapsed.
+			return self
+		}
+
+		if self.principal.eq(&0i16) {
+			return Self { principal: self.principal, last_update: current_block_number }
+		}
+
+		let elapsed_blocks =
+			current_block_number.checked_sub(&self.last_update).unwrap_or_else(|| {
+				// This should never be the case on a sound blockchain setup, but we really
+				// never want to panic here.
+				warn!(
+					target: LOG,
+					"last block {:?} bigger than current {:?}, defaulting to elapsed_blocks=0",
+					self.last_update,
+					current_block_number
+				);
+				0u32.into()
+			});
+
+		let elapsed_u32: u32 = elapsed_blocks.try_into().unwrap_or_else(|_| {
+			// 698 years with 6 seconds block time.
+			trace!(target: LOG, "elapsed_blocks > u32::MAX, defaulting to u32::MAX");
+			u32::MAX
+		});
+
+		let demurrage_factor = demurrage_factor(demurrage_per_block, elapsed_u32);
+
+		let principal = self
+			.principal
+			.checked_mul(demurrage_factor)
+			.expect("demurrage_factor [0,1), hence can't overflow; qed");
+
+		Self { principal, last_update: current_block_number }
+	}
+}
+
+/// e^(-demurrage_per_block * elapsed_blocks) within [0,1).
+///
+/// It will take the absolute value of the `demurrage_per_block` if it is negative.
```

Comment:
It is a detail but to me it seems more logical to use `max(demurrage_per_block, 0)` than `demurrage_per_block.abs()`

Reply:
For me the idea was that maybe `e^-demurrage*block` is unclear and that they accidentally include the `-` into the demurrage.

Reply:
ok, also a valid point :)